### DISCUSSION TITLE
'chezmoi cd' and Subshells

### DISCUSSION THREAD

**Discussion by cadenkriese** [twpayne/chezmoi#3403] [Q&A] [answered]
I use `chezmoi cd` as an easy way to get to my Chezmoi directory in my terminal instead of typing out the whole path. One user flow I usually go through that doesn't work is when I'm modifying something and I `chezmoi cd`, edit it, run `chezmoi apply`, check the changes in `cd ~`, then when I try to `chezmoi cd` again it errors because I'm already in a subshell. I looked at [the documentation](https://www.chezmoi.io/reference/commands/cd/), which explains why the subshell feature exists: to set environment variables for scripts.

My question is, is there anything Chezmoi can do to help with my workflow? I know I could make an alias to cd to that directory or use an application like [z](https://github.com/skywind3000/z.lua), but I want to see if there's anything in chezmoi first.

**Comment by halostatue** [comment, marked as answer]
The only real option is to make a shell function. The environment variable setting is a *side-effect*, but not the reason that it exists. Without a shell function, `chezmoi cd` has to use a subshell in order to present you with the changed directory, because otherwise `chezmoi` will change the directory and exit — which will leave you in the same directory you started from, since `chdir` in a subprocess only affects that subprocess.

An alias will *not* do what you want, you need a shell function:

```bash
ccd() {
  cd "$(chezmoi source-path "$@")"
}
```

Within your current described workflow, I can suggest a few improvements:

1. After using `cd ~` from the `chezmoi cd` subshell, use `cd -`.
2. Use `pushd ~` and `popd` when done (I think `pushd` and `popd` exist in bash if that's what you use).
3. Use `chezmoi diff` instead of `chezmoi apply` and then checking, if you know your expected output well enough to see what the difference would be between what is present and what would be written.
4. Use `chezmoi merge-all` to review larger sets of changes if you edit the target files frequently (this is my main operation).

**Reply by cadenkriese** [reply]
Well worded and extremely helpful. Thank you!
